### DISCUSSION TITLE
Add extra location to nginx-s3-gateway image

### DISCUSSION THREAD

**Discussion by MuriloCouceiro** [nginxinc/nginx-s3-gateway#189] [Q&A] [answered]
Hello, I'm trying to add extra locations to the S3 gateway that map a local folder in order to upload or download files.

When trying to add these locations to the nginxinc/nginx-s3-gateway image configs, these are overwritten with the original image ones.

I'm trying with the following file structure and Dockerfile, however, it is not working as described above. Can you tell me if I'm doing anything wrong? 

Dockerfile
```

FROM nginxinc/nginx-s3-gateway

ENV S3_BUCKET_NAME "bucket"
ENV AWS_ACCESS_KEY_ID "aaaa"
ENV AWS_SECRET_ACCESS_KEY "aaaa"
ENV S3_SERVER "aaaaa"
ENV S3_SERVER_PORT "80"
ENV S3_SERVER_PROTO "http"
ENV S3_REGION "us-east-1"
ENV S3_STYLE "path"
ENV DEBUG "false"
ENV AWS_SIGS_VERSION "4"
ENV ALLOW_DIRECTORY_LIST "true"
ENV PROVIDE_INDEX_PAGE "true"
ENV NGINX_ENTRYPOINT_QUIET_LOGS "true"

COPY etc/nginx/conf.d /etc/nginx/conf.d 

COPY . /var/uploads
```

s3_server.conf example 

```
location /local/health {
    return 200;
}
```

File Structure 
![image](https://github.com/nginxinc/nginx-s3-gateway/assets/72820316/29651c4c-c1e6-4384-b342-8abbadac3218)

Thanks in advance

**Comment by 4141done** [comment]
Hello, thanks for reaching out and sorry to hear you are having problems modifying the image. There are two strategies you can use to modify the NGINX configuration.

I created a [gist](https://gist.github.com/4141done/de938b378cacb9a64b1c3ff08d9e4cf6) showing my understanding of your question.

The directory structure is as follows:
```
.
├── Dockerfile
├── etc
│   └── nginx
│       └── conf.d
│           ├── gateway
│           │   └── s3_server.conf
│           └── modifications.conf
└── images
    └── whale.jpg
```

Then, running
```bash
docker build -t my_image:1 .
```
to build the image.

And 
```bash
docker run -it custom_s3:1 /bin/bash
```

Shows the modified files in the correct locations.  They will be loaded by [this statement](https://github.com/nginxinc/nginx-s3-gateway/blob/a95a6f3479ff7fd128fa9e2a42b859887a457394/common/etc/nginx/nginx.conf#L57) in the main configuration.


## Another method
Alternately, you could build your docker image using this repository checked out locally, make your modifications to 
[s3_server.conf.template](https://github.com/nginxinc/nginx-s3-gateway/blob/master/common/etc/nginx/templates/gateway/s3_server.conf.template) then add your additional files (the ones you want to add to `var/uploads`) in the Dockerfile that corresponds to your build (open source one is [here](https://github.com/nginxinc/nginx-s3-gateway/blob/master/Dockerfile.oss). You'd place it with the other `COPY` statements.

Then build:
```
docker build -f Dockerfile.oss\
      --tag nginx-s3-gateway --tag nginx-s3-gateway:oss.
```

Hope this helps.  Please let me know if anything is not clear or if I misunderstood your problem.

**Reply by MuriloCouceiro** [reply]
Thank you for the quick response. 

I'm trying the first option and in fact, I do see the modified files. However, I'll ask for your help again on how to start the nginx server. 

With the first option, if I start the container without the bin/bash entrypoint  `docker run my_image:1` I get the following error:

![image](https://github.com/nginxinc/nginx-s3-gateway/assets/72820316/8c4e8f95-9c2d-438c-93c4-4d46a31b4d40)

**Reply by 4141done** [reply]
Apologies, I gave you incorrect advice (I'm still learning as a maintainer of this project).

Any files you include in `/etc/nginx/conf.d` will be included in the `http` context.  This can be helpful, but in your case you don't want to create a whole new [virtual server](https://www.nginx.com/resources/wiki/start/topics/examples/server_blocks/).

However, the specific file [/etc/nginx/conf.d/gateway/s3_server.conf](https://github.com/nginxinc/nginx-s3-gateway/blob/master/common/etc/nginx/templates/gateway/s3_server.conf.template) file is included in the main `server` block for the s3 gateway (see [here](https://github.com/nginxinc/nginx-s3-gateway/blob/a95a6f3479ff7fd128fa9e2a42b859887a457394/common/etc/nginx/templates/default.conf.template#L345)) so you can add `location` directives directly there.

That means if you want to add an extra location block you'd want to override that file.  The [gist](https://gist.github.com/4141done/de938b378cacb9a64b1c3ff08d9e4cf6) I originally posted has an example of this and I've confirmed that building in this way does override the file.

```bash
# Start the container and inspect the templated files
docker run -p 80:80 -it my_image:1 /bin/bash
root@83e7f37acf15:/# cat /etc/nginx/conf.d/gateway/s3_server.conf
location /local/health {
    return 200;
}

location /foo/bar {
    return  200 "this is foo/bar";
}

root@83e7f37acf15:/# exit



# Start the container using the normal command
docker run -p 80:80 my_image:1
S3 Backend Environment
Access Key ID: aaaa
Origin: http://bucket.google.com:80
Region: us-east-1
Addressing Style: path
AWS Signatures Version: v4
DNS Resolvers:  192.168.65.7
Directory Listing Enabled: true
Directory Listing Path Prefix:
Provide Index Pages Enabled: true
Append slash for directory enabled:
Stripping the following headers from responses: x-amz-;
CORS Enabled: 0
2023/11/29 22:44:04 [notice] 1#1: using the "epoll" event method
2023/11/29 22:44:04 [notice] 1#1: nginx/1.25.3
2023/11/29 22:44:04 [notice] 1#1: built by gcc 12.2.0 (Debian 12.2.0-14)
2023/11/29 22:44:04 [notice] 1#1: OS: Linux 6.4.16-linuxkit
2023/11/29 22:44:04 [notice] 1#1: getrlimit(RLIMIT_NOFILE): 1048576:1048576
2023/11/29 22:44:04 [notice] 1#1: start worker processes
2023/11/29 22:44:04 [notice] 1#1: start worker process 206
2023/11/29 22:44:04 [notice] 1#1: start cache manager process 208
2023/11/29 22:44:04 [notice] 1#1: start cache loader process 210
```

**Summary**
> Add your changes to a file called `/etc/nginx/conf.d/gateway/s3_server.conf` and copy it into the image on build as shown in the gist.

**Reply by MuriloCouceiro** [reply]
Thanks again for the answer! 

Without that modification.conf file, I think it is the same as I initially tested this. I've tried once again and if running the container without the entrypoint, somehow the files are overwritten with the original ones. 

Here are some prints with red highlighting showing the behaviour. First I execute the commands you've shown here and after the server is running, I cannot reach the desired location. 

![dockerrun](https://github.com/nginxinc/nginx-s3-gateway/assets/72820316/c5e02862-42ca-4967-81dc-882de6f50b1f)

When looking into the s3_server.conf file it turns out that it does not contain the changes I've made (2nd image).
 
![exec](https://github.com/nginxinc/nginx-s3-gateway/assets/72820316/1292689a-50e5-41c5-846c-43197d32dfd7)

**Reply by MuriloCouceiro** [reply]
@4141done any news on this?

**Reply by 4141done** [reply, marked as answer]
Hi @MuriloCouceiro ! Sorry for the delay.  You are correct that in my example gist the contents of `s3_server.conf` are being overwritten when the normal docker entrypoint is used.

The reason for this is that the gateway project takes the files in `/etc/nginx/templates`, runs `envsubst` on them at startup, and outputs the files to `/etc/nginx/conf.d`.

In my gist, I mistakenly put the file in `/etc/nginx/conf.d/gateway` while was overwritten by that process. Apologies for the bad initial answer.  I am still learning as a maintainer on this project 🙏 .

So here's a new solution that I've tested.

1. Place your modified file in `etc/nginx/templates/gateway/s3_server.conf.template` (it can be anywhere  but I find it helps to match the folder structure for the docker `COPY` command).
```
.
├── Dockerfile
├── etc
│   └── nginx
│       └── templates
│           └── gateway
│               └── s3_server.conf.template
```
The `.template` part of the name is important. Because the project includes this file by name the entire name must be exactly `s3_server.conf.template`
2. Adjust the `COPY` command in the `Dockerfile` to: `COPY etc/nginx/templates /etc/nginx/templates`
3. Build and run the image.

Here are my results.
The `s3_server.conf.template` contains:
```nginx
location /local/health {
    return 200;
}

location /foo/bar {
    return  200 "this is foo/bar";
}
```

Starting the docker container and the following logs show the `location` defined in my custom `s3_server.conf.template` being served:
```
docker run --platform linux/amd64 -it -p 80:80 my_image:1
S3 Backend Environment
Access Key ID: aaaa
Origin: http://bucket.google.com:80
Region: us-east-1
Addressing Style: path
AWS Signatures Version: v4
DNS Resolvers:  192.168.65.7
Directory Listing Enabled: true
Directory Listing Path Prefix:
Provide Index Pages Enabled: true
Append slash for directory enabled:
Stripping the following headers from responses: x-amz-;
CORS Enabled: 0
2023/12/09 00:26:15 [notice] 1#1: using the "epoll" event method
2023/12/09 00:26:15 [notice] 1#1: nginx/1.25.3
2023/12/09 00:26:15 [notice] 1#1: built by gcc 12.2.0 (Debian 12.2.0-14)
2023/12/09 00:26:15 [notice] 1#1: OS: Linux 6.4.16-linuxkit
2023/12/09 00:26:15 [notice] 1#1: getrlimit(RLIMIT_NOFILE): 1048576:1048576
2023/12/09 00:26:15 [notice] 1#1: start worker processes
2023/12/09 00:26:15 [notice] 1#1: start worker process 75
2023/12/09 00:26:15 [notice] 1#1: start cache manager process 76
2023/12/09 00:26:15 [notice] 1#1: start cache loader process 77
192.168.65.1 - - [09/Dec/2023:00:26:18 +0000] "GET /foo/bar HTTP/1.1" 200 15 "-" "curl/8.1.2" "-"
2023/12/09 00:26:18 [info] 75#75: *1 client 192.168.65.1 closed keepalive connection
```

Testing requests to the new endpoint:
```
curl localhost/foo/bar
this is foo/bar
```

Let me know if that works for you! If it does I'll be adding some guidance to the README to make this method of modifying the project clear for future users.

**Reply by MuriloCouceiro** [reply]
Hello, @4141done! This works for me as well! 
Thank you so much for the help!
